src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/step/forcemerge/WaitForForceMergeStep.kt

```diff
@@ -0,0 +1,145 @@
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.step.Step
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.service.ClusterService
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.action.admin.indices.stats.IndicesStatsRequest
+import org.elasticsearch.action.admin.indices.stats.IndicesStatsResponse
+import org.elasticsearch.rest.RestStatus
+import java.time.Duration
+import java.time.Instant
+
+class WaitForForceMergeStep(
+    val clusterService: ClusterService,
+    val client: Client,
+    val config: ForceMergeActionConfig,
+    managedIndexMetaData: ManagedIndexMetaData
+) : Step(name, managedIndexMetaData) {
+
+    private val logger = LogManager.getLogger(javaClass)
+    private var stepStatus = StepStatus.STARTING
+    private var info: Map<String, Any>? = null
+
+    @Suppress("TooGenericExceptionCaught") // TODO see if we can refactor to catch GenericException in Runner.
+    override suspend fun execute() {
+        val indexName = managedIndexMetaData.index
+
+        logger.info("Waiting for force merge to complete on [$indexName]")
+
+        // Retrieve maxNumSegments value from ActionProperties. If ActionProperties is null, update failed info and return early.
+        val maxNumSegments = getMaxNumSegments() ?: return
+
+        // Get the number of shards with a segment count greater than maxNumSegments, meaning they are still merging
+        val shardsStillMergingSegments = getShardsStillMergingSegments(indexName, maxNumSegments)
+        // If shardsStillMergingSegments is null, failed info has already been updated and can return early
+        shardsStillMergingSegments ?: return
+
+        // If there are no longer shardsStillMergingSegments, then the force merge has completed
+        if (shardsStillMergingSegments == 0) {
+            logger.info("Force merge completed on [$indexName]")
+
+            stepStatus = StepStatus.COMPLETED
+            info = mapOf("message" to "Force merge completed")
+        } else {
+            /*
+             * If there are still shards with segments merging then no action is taken and the step will be reevaluated
+             * on the next run of the ManagedIndexRunner.
+             *
+             * However, if an internal timeout of 12 hours is reached, the force_merge action will be marked as failed.
+             * This is for cases where the force merge stops silently, such as when nodes relocate during a force
+             * merge operation. Without a timeout, the segment count would stop going down as merging would no longer
+             * occur and the managed index would become stuck in this action.
+             */
+            val timeWaitingForForceMerge: Duration = Duration.between(getStepStartTime(), Instant.now())
+            if (timeWaitingForForceMerge.toSeconds() > FORCE_MERGE_TIMEOUT_IN_SECONDS) {
```

Comment:
What happens if we have an internal force merge timeout of 12 hours and then they set a timeout on the force merge action as 48 hours. Should we respect theirs or ours?

Reply:
Changed this to favor ```ActionTimeout``` if given and use our internal 12 hour timeout if nothing was provided.